### PR TITLE
Introduce tests and clean

### PR DESCRIPTION
**Pull request by mwhipple**
This includes some loose ends that should be tied up in subsequent PRs: this one focuses on specifying the current behavior while the purpose is to support some of the extensions which can now be added (such as a pluggable versioning strategy)

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -0,0 +1,10 @@
+include bml
+
+DOCKER = $(call required-command,docker)
```
Comment (jackterranova) [inline Makefile]: Is this used?

Comment (mwhipple) [reply Makefile]: Ahh, not yet...that's for the TODO item listed below but It should be removed before merging

**Review by jackterranova — APPROVED** [review]
Maybe add a blurb about installing `bats`.

Some of the tests are stalling and a CR seems to get them moving.   🤷‍♂ 

And some tests do fail.

Is a particular bash version required?

```
✗ Reflect the release in the git log.
   (in test file test/release/post_release.bats, line 34)
     `mapfile -t logs < <(git log HEAD~2..HEAD --oneline --no-decorate | cut -d ' ' -f2-)' failed with status 127
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/4941/logs_origin/
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/4941/logs/.git/
   [master (root-commit) f4c6482] Sync to origin
    2 files changed, 471 insertions(+)
    create mode 100644 build.properties
    create mode 100755 release
   To file:///var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/4941/logs_origin
    * [new branch]      master -> master
   Branch master set up to track remote branch master from origin.
   /var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/bats.4756.src: line 34: mapfile: command not found
```

I am not sure why these are failing...

```
✗ When untracked files exist, output a fitting message.
   (in test file test/release/unsynched.bats, line 18)
     `[[ "${output}" =~ Modified ]]' failed with status 128
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6692/untracked_origin/
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6692/untracked/.git/
   [master (root-commit) 8b53d0b] Sync to origin
    1 file changed, 470 insertions(+)
    create mode 100755 release
   To file:///var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6692/untracked_origin
    * [new branch]      master -> master
   Branch master set up to track remote branch master from origin.
```

```
✗ With unstaged modifications, output a fitting message.
   (in test file test/release/unsynched.bats, line 54)
     `[[ "${output}" =~ Modified ]]' failed
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6863/unstaged_origin/
   Initialized empty Git repository in /private/var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6863/unstaged/.git/
   [master (root-commit) adc2c77] Sync to origin
    2 files changed, 470 insertions(+)
    create mode 100644 build.properties
    create mode 100755 release
   To file:///var/folders/sd/tfh9bwk161l9rbqt2kvcvv440000gp/T/6863/unstaged_origin
    * [new branch]      master -> master
   Branch master set up to track remote branch master from origin.
```

**Comment by mwhipple** [comment]
A new bash version is likely required to run the tests, yes. With testing I normally use new bash constructs without worrying about portability. That should be dockerized to avoid the issues and also so Travis can run the tests easily (hence the beginnings of docker in the Makefile).

**Pull request merged**